### PR TITLE
Replace `OrderedMultiIndex` with `Combination`

### PR DESCRIPTION
**Pull request by mjrodgers**
There are a few things that need to be adapted to make this replacement, namely we need

- [x]  a ranking method to get the index of a given `Combination` C in `Combinations(n,k)`
- [x] an unranking method to get the ith `Combination` C appearing in `Combinations(n,k)` directly
- [x] adapt `_wedge(a,b)` for a and b combinations that returns `sign, c` where, if a and b contain a common element, `sign = 0`, or else `sign` is +/-1 depending on the sign of the required to order `vcat(a,b)`, and `c` is the new `Combination` obtained through this
- [x] make a version of `_wedge` that takes a vector of `Combination` (I don't think this is used anywhere though?)
- [x] make some optimizations of the above methods

Once this is complete, we can replace the use of `OrderedMultiIndex` in
- [x]  `src/Modules/ExteriorPowers`, and 
- [x] `experimental/DoubleAndHyperComplexes`
- [x] `experimental/Schemes/src/Resolution_structure.jl`

### CODE REVIEW

**Comment by mjrodgers** [comment]
I have been experimenting with "ranking" and "unranking" algorithms for subsets (these are for determining the index of a given k-set in the lexicographic ordering of all such sets, and conversely determining a k-set for a given index).

I was able to get a significant improvement to the ranking algorithm (`linear_index`), it runs in about half the time for small values of `n` with bigger improvements as the size increases. ~~For the unranking algorithm, my new version seems to have very similar performance to the older `OrderedMultiIndex` version, but is no longer done recursively. It seems like getting better performance for this is difficult when using lexicographic ordering.~~

EDIT: After revisiting the unranking algorithm and fixing a few small mistakes I had made, there is actually about a 10x speed increase at least for (n,k) = (27,7).

File: experimental/DoubleAndHyperComplexes/src/Objects/Methods.jl [hunk 1]
```diff
@@ -2,10 +2,10 @@
 @doc raw"""
     total_complex(D::AbsDoubleComplexOfMorphisms)
 
-Construct the total complex of the double complex `D`. 
+Construct the total complex of the double complex `D`.
```
Comment (lgoettgens) [inline experimental/DoubleAndHyperComplexes/src/Objects/Methods.jl:5]: This PR contains a lot of whitespace changes like this here. Could you try to drop these kinds of changes from the PR to make it smaller and thus easier to review (and reduce the chance of conflicts)?

**Comment by lgoettgens** [comment]
There seems to be one more use of `OrderedMultiIndex` in https://github.com/oscar-system/Oscar.jl/blob/0a7fd981043c579d8ba7feba381e69d8e4b91774/experimental/Schemes/src/Resolution_structure.jl#L875. Do you want to attempt replacing that with some combinations functions here or in a follow-up PR?
The same question would be for deleting the src and test files for `OrderedMultiIndex`.

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 2]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+#  Ranking / unranking combinations
+#
+################################################################################
+
+# For a combination C containing k elements from 1:n,
+# return the index i so that C occurs as the ith element in the
+# iteration over all such combinations (which are lexicographically ordered).
+# NOTE We use the method described in J. Liebehenschel - Ranking and unranking of lexicographically ordered words: an average-case analysis. Journal of Automata, Languages, and Combinatorics (1997).
+# NOTE we have added an extra argument, since this info is not stored with
+# a Combination as it is with an OrderedMultiIndex
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: maybe convert this to a docstring?

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 3]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+function linear_index(C::Combination, n::IntegerUnion)
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl:152]: I think this function assumes that all entries in `C` are bounded by `n`, right? If yes, this should be at least mentioned in the docstring. maybe even checked?

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 4]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+    r -= binomial(n-C[i], k-i+1)
+    i += 1
+    k-i+1 == n-C[i-1] && return r
+  end
+  return r
+end
+
+# Return the rth combination in the iteration over combinations(n,k)
+# NOTE We use the method described in J. Liebehenschel - Ranking and unranking of lexicographically ordered words: an average-case analysis. Journal of Automata, Languages, and Combinatorics (1997).
+# NOTE arguments are reordered to match combinatorial conventions
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]:
```suggestion
```
IMO not relevant in the codebase as one won't have any context from where this was taken etc.

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 5]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+function combination(n::Int, k::Int, r::Int)
+  (r < 1 || r > binomial(n, k)) && error("index out of range")
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]:
```suggestion
  @req 1 <= r <= binomial(n, k) "index out of range"
```

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 6]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+      j += 1
+    end
+  end
+
+  return result, sign
+end
+
+
+# For two combinations a = [i₁, i₂, …, iₚ] and b = [j₁, j₂, …, jᵣ]
+# the result is a pair `(sign, c)` with a a new combination,
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]:
```suggestion
# the result is a pair `(sign, c)` with `c` a new combination,
```

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 7]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+# and `sign` either 0 in the case that
+# iₖ = jₗ for some k and l, or ±1 depending on the number of transpositions
+# needed to put [i₁, …, iₚ, j₁, …, jᵣ] into a strictly increasing order
+# to produce `c`.
+function _wedge(a::Combination{T}, b::Combination{T}) where {T}
+  # if combinations are not disjoint, return 0
+  any(in(b), a) && return 0, a
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]:
```suggestion
  isdisjoint(a, b) || return 0, a
```

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 8]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+
+  c, sign = merge_sorted_with_sign(data(a), data(b))
+  return sign, Combination(c)
+end
+
+
+# # This could be optimized, but don't think it is currently used anywhere so leaving it for now.
+function _wedge(a::Vector{T}) where {T <: Combination}
+  isempty(a) && error("list must not be empty")
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]:
```suggestion
  @req !isempty(a) "list must not be empty"
```

File: test/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 9]
```diff
@@ -42,4 +42,43 @@
   @test allunique(v)
 
   @test collect(combinations(0, 0)) == [Int[]]
+
+  @testset "ranking/unranking" begin
+    n = 5
+    k = 3
```
Comment (lgoettgens) [inline test/Combinatorics/EnumerativeCombinatorics/combinations.jl:52]: could you also add some ranking/unranking tests for edgecases like `n = k = 0`, `n = k > 0`, `n > k = 1`? I just got bitten for not having a tests for such a case (x-ref https://github.com/oscar-system/Oscar.jl/pull/4878)

File: experimental/DoubleAndHyperComplexes/src/Objects/Methods.jl (hunk 1 above)
Comment (mjrodgers) [reply experimental/DoubleAndHyperComplexes/src/Objects/Methods.jl:5]: I can try to avoid this in the future. My editor strips this kind of whitespace on save (I can turn this setting off), it wasn't an intentional formatting step I made - so for this pull request it's sort of hopelessly mixed in to my commits.

**Comment by mjrodgers** [comment]
> There seems to be one more use of `OrderedMultiIndex` in
> . Do you want to attempt replacing that with some combinations functions here or in a follow-up PR?
> The same question would be for deleting the src and test files for `OrderedMultiIndex`.

I can replace this use of `OrderedMultiIndex`, though maybe I would delete the original `OrderedMultiIndex` src and test files in a follow-up PR.

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl (hunk 3 above)
Comment (mjrodgers) [reply src/Combinatorics/EnumerativeCombinatorics/combinations.jl:152]: Yeah, for `OrderedMultiIndex` the bound `n` is stored as part of the `struct`, which is not the case for a `Combination` so it made sense to include this as an extra argument. A check is pretty easy to implement here, so I will add one (+ a docstring).

File: experimental/DoubleAndHyperComplexes/src/Objects/Methods.jl (hunk 1 above)
Comment (mjrodgers) [reply experimental/DoubleAndHyperComplexes/src/Objects/Methods.jl:5]: (for reviewing, you can click the settings dropdown in GitHub and have it `ignore whitespace`, then those changes won't show up in the diff)

File: experimental/DoubleAndHyperComplexes/src/Objects/induced_ENC.jl [hunk 10]
```diff
@@ -22,8 +22,8 @@ function (fac::InducedENCChainFactory)(self::AbsHyperComplex, ind::Tuple)
   k + i - 1 > ngens(I0) && return sub(enc[i], elem_type(enc[i])[])[1]
   amb_ext_power = _exterior_power(enc, i)
   new_gens = elem_type(amb_ext_power)[]
-  for a in OrderedMultiIndexSet(k+i-1, ngens(I0))
-    push!(new_gens, wedge([repres(gen(I0, i)) for i in indices(a)]; parent=amb_ext_power))
+  for a in combinations(ngens(IO), k+i-1)
```
Comment (lgoettgens) [inline experimental/DoubleAndHyperComplexes/src/Objects/induced_ENC.jl]:
```suggestion
  for a in combinations(ngens(I0), k+i-1)
```

File: experimental/DoubleAndHyperComplexes/src/Objects/Methods.jl [hunk 11]
```diff
@@ -359,15 +359,13 @@ function _contract(
...
   result = zero(parent)
   n = ngens(F0)
-  for (i, ind) in enumerate(OrderedMultiIndexSet(p, n))
+  for (i, ind) in enumerate(combinations(n, p))
     is_zero(v[i]) && continue
     for j in 1:p
-      I = deleteat!(copy(indices(ind)), j)
-      new_ind = OrderedMultiIndex(I, n)
-      result = result + (-1)^j * v[i] * phi[ind[j]] * parent[linear_index(new_ind)]
+      I = deleteat!(copy(data(ind)), j)
```
Comment (fingolfin) [inline experimental/DoubleAndHyperComplexes/src/Objects/Methods.jl:365]: not a complaint, just a musing: why doesn't Julia have `deletat(data(ind),j)` (i.e. without the `!`)? *sigh*

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 12]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+################################################################################
+#
+#  Ranking / unranking combinations
+#
+################################################################################
+
+# For a combination C containing k elements from 1:n,
+# return the index i so that C occurs as the ith element in the
+# iteration over all such combinations (which are lexicographically ordered).
+# NOTE We use the method described in J. Liebehenschel - Ranking and unranking of lexicographically ordered words: an average-case analysis. Journal of Automata, Languages, and Combinatorics (1997).
```
Comment (fingolfin) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: Maybe put this into the bibliography?

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 13]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+#
+#   # in case of a double index return zero
+#   any(in(b), a) && return 0, data(a)
+#
+#   result, sign = merge_sorted_with_sign(data(a),data(b))
+#
+#   return sign, result_indices
+# end
+
+
```
Comment (fingolfin) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]:
```suggestion
```

@HechtiDerLachs do you need this resp. expect to need this? If not we should just delete it.

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 14]
```diff
@@ -92,24 +92,155 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+  @req !isempty(a) "list must not be empty"
+  isone(length(a)) && return 1, first(a)
+  k = div(length(a), 2)
+  b = a[1:k]
+  c = a[k+1:end]
+  sign_b, comb_b = _wedge(b)
+  sign_c, comb_c = _wedge(c)
+  sign, comb = _wedge(comb_b, comb_c)
+  return sign * sign_b * sign_c, comb
 end
```
Comment (fingolfin) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: If we don't need it, remove it? Unless @HechtiDerLachs foresees a future need for it?
```suggestion
```

File: experimental/DoubleAndHyperComplexes/src/Objects/induced_ENC.jl [hunk 15]
```diff
@@ -22,8 +22,8 @@ function (fac::InducedENCChainFactory)(self::AbsHyperComplex, ind::Tuple)
   k + i - 1 > ngens(I0) && return sub(enc[i], elem_type(enc[i])[])[1]
   amb_ext_power = _exterior_power(enc, i)
   new_gens = elem_type(amb_ext_power)[]
-  for a in OrderedMultiIndexSet(k+i-1, ngens(I0))
-    push!(new_gens, wedge([repres(gen(I0, i)) for i in indices(a)]; parent=amb_ext_power))
+  for a in combinations(ngens(I0), k+i-1)
+    push!(new_gens, wedge([repres(gen(I0, i)) for i in a]; parent=amb_ext_power))
```
Comment (lgoettgens) [inline experimental/DoubleAndHyperComplexes/src/Objects/induced_ENC.jl]:
```suggestion
  for as in combinations(gens(I0), k+i-1)
    push!(new_gens, wedge([repres(a) for a in as]; parent=amb_ext_power))
```
shouldn't this work as well here and saves some allocations?

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 16]
```diff
@@ -92,24 +92,184 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+#
+################################################################################
+
+
+@doc raw"""
+    linear_index(C::Combination, n::IntegerUnion)
+
+For a combination C containing k elements from 1:n,
+return the index i so that C occurs as the ith element in the
+iteration over all such combinations (in lexicographic order).
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]:
```suggestion
For a combination `C` containing `k` elements from `1:n`,
return the index `i` so that `C` occurs as the `i`th element in the
iteration over all such combinations (in lexicographic order).
```

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 17]
```diff
@@ -92,24 +92,184 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+    i += 1
+    k-i+1 == n-C[i-1] && return r
+  end
+  return r
+end
+
+@doc raw"""
+    combination(n::Int, k::Int, r::Int)
+
+Return the rth combination in the iteration over combinations(n,k).
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]:
```suggestion
Return the `r`th combination in the iteration over `combinations(n,k)`.
```

File: src/Modules/ExteriorPowers/FreeModules.jl [hunk 18]
```diff
@@ -20,12 +20,12 @@ function exterior_power(F::FreeMod, p::Int; cached::Bool=true)
...
 
-  # In case F was graded, we have to take an extra detour. 
+  # In case F was graded, we have to take an extra detour.
   result = if is_graded(F)
     G = grading_group(F)
     weights = elem_type(G)[]
-    for ind in OrderedMultiIndexSet(p, n)
-      push!(weights, sum(_degree_fast(F[i]) for i in indices(ind); init=zero(G)))
+    for ind in combinations(n, p)
+      push!(weights, sum(_degree_fast(F[i]) for i in ind; init=zero(G)))
```
Comment (lgoettgens) [inline src/Modules/ExteriorPowers/FreeModules.jl]:
```suggestion
    for fs in combinations(F, p)
      push!(weights, sum(_degree_fast(f) for f in fs; init=zero(G)))
```

File: src/Modules/ExteriorPowers/FreeModules.jl [hunk 19]
```diff
@@ -47,9 +47,9 @@ function exterior_power(F::FreeMod, p::Int; cached::Bool=true)
     @req parent(u) === result "element does not belong to the correct module"
     k = findfirst(==(u), gens(result))
     @req !isnothing(k) "element must be a generator of the module"
-    ind = ordered_multi_index(k, p, n)
+    ind = combination(n, p, k)
     e = gens(F)
-    return Tuple(e[i] for i in indices(ind))
+    return Tuple(e[i] for i in ind)
```
Comment (lgoettgens) [inline src/Modules/ExteriorPowers/FreeModules.jl]:
```suggestion
    return Tuple(gen(F, i) for i in ind)
```
this only accesses a subset of the gens, so it is better not not allocate all of them but only the ones we actually need.

File: src/Modules/ExteriorPowers/FreeModules.jl [hunk 20]
```diff
@@ -47,9 +47,9 @@ function exterior_power(F::FreeMod, p::Int; cached::Bool=true)
     @req parent(u) === result "element does not belong to the correct module"
     k = findfirst(==(u), gens(result))
     @req !isnothing(k) "element must be a generator of the module"
-    ind = ordered_multi_index(k, p, n)
+    ind = combination(n, p, k)
     e = gens(F)
```
Comment (lgoettgens) [inline src/Modules/ExteriorPowers/FreeModules.jl]:
```suggestion
```
no longer needed when next suggestion is applied

File: src/Modules/ExteriorPowers/Generic.jl [hunk 21]
```diff
@@ -178,7 +178,7 @@ function induced_map_on_exterior_power(phi::FreeModuleHom{<:FreeMod, <:FreeMod,
   is_zero(p) && return hom(domain, codomain, gens(codomain); check=false) # Isomorphism of R^1
 
   imgs = phi.(gens(F))
-  img_gens = [wedge(imgs[indices(ind)], parent=codomain) for ind in OrderedMultiIndexSet(p, m)]
+  img_gens = [wedge(imgs[data(ind)], parent=codomain) for ind in combinations(m, p)]
```
Comment (lgoettgens) [inline src/Modules/ExteriorPowers/Generic.jl]:
```suggestion
  img_gens = [wedge(data(c), parent=codomain) for c in combinations(imgs, p)]
```

File: src/Modules/ExteriorPowers/SubQuo.jl [hunk 22]
```diff
@@ -32,9 +32,9 @@ function exterior_power(M::SubquoModule, p::Int; cached::Bool=true)
     @req parent(u) === result "element does not belong to the correct module"
     k = findfirst(==(u), gens(result))
     @req !isnothing(k) "element must be a generator of the module"
-    ind = ordered_multi_index(k, p, n)
+    ind = combination(n, p, k)
     e = gens(M)
```
Comment (lgoettgens) [inline src/Modules/ExteriorPowers/SubQuo.jl]:
```suggestion
```
see above

File: src/Modules/ExteriorPowers/SubQuo.jl [hunk 23]
```diff
@@ -32,9 +32,9 @@ function exterior_power(M::SubquoModule, p::Int; cached::Bool=true)
...
-    return Tuple(e[i] for i in indices(ind))
+    return Tuple(e[i] for i in ind)
```
Comment (lgoettgens) [inline src/Modules/ExteriorPowers/SubQuo.jl]:
```suggestion
    return Tuple(gen(M, i) for i in ind)
```

File: src/exports.jl [hunk 24]
```diff
@@ -1079,6 +1080,7 @@ export lifted_numerator
 export lineality_dim
 export lineality_space
 export linear_characters
+export linear_index
```
Comment (lgoettgens) [inline src/exports.jl]: I am not convinced by this name (at least not for exporting it under that name). Note that there already exists `LinearIndices` in julia base, which does something completely different (allowing arrays of arbitrary dimension being indexed by a single Int).

Comment (mjrodgers) [reply src/exports.jl]: I'm on the fence about it.... To me, the set `C = combinations(n,k)` can be made into an `AbstractVector` that is linearly indexed in the sense of Julia `Base` (that is, `IndexStyle` is `IndexLinear`), and `linear_index` will give the index of an element `c` of `C`, so it makes at least a little bit of sense.

I really just carried over the name from `OrderedIndexSet`, I'm happy to change it to something else if it makes sense. But it's worth noting that in MAGMA this would be `Index(C, c)` to obtain the position of `c`.

File: experimental/DoubleAndHyperComplexes/src/Objects/induced_ENC.jl (hunk 15 above)
Comment (mjrodgers) [reply experimental/DoubleAndHyperComplexes/src/Objects/induced_ENC.jl]: thanks for all of these changes, I guess I just sort of naively changed a lot of the `OrderedMultiSet` to `Combination` without paying attention to some of this added functionality we have (taking subsets or arbitrary objects instead of just `1:n`).

**Comment by mjrodgers** [comment]
OK, seems mostly sorted aside from adding tests for edge cases, deciding naming of `linear_index`, and removing `_wedge(V::Vector)` if there is no planned future use. I'll take care of these last things before triage tomorrow.

File: src/exports.jl (hunk 24 above)
Comment (fingolfin) [reply src/exports.jl]: To me the natural interface would be this:
- instead of `combination(n, k, r)`, provide a `getindex` method for `Combinations` which then allows us to do `combinations(n, k)[r]` i.e the `Combinations` object really is treated like a vector
- instead of `linear_index`... provide whatever one uses to "search" in a `Vector`  (actually I am not aware of a great API for that... the closes I can think of is `indexin`, but the result of e.g. `indexin(17,[1,17,3])` is a 0-dimensional array containing the index (or `nothing` if not found) which is a bit odd

Comment (mjrodgers) [reply src/exports.jl]: The 0-dimensional array is a bit weird... it looks like (from the documentation) julia is intending `indexin(a,b)` to have `a` be a vector of elements that may or may not be in `b`, in this case we get a 1-dimensional array with the same length as `a` (with the corresponding indices or `nothing`).

~~Does it make more sense to use `findfirst` instead of `indexin`?~~ No, this doesn't really look like it makes sense.

**Comment by fingolfin** [comment]
@mjrodgers new test failures:
```
┌ Error: doctest failure in src/Combinatorics/EnumerativeCombinatorics/combinations.jl:9-19
│ 
│ ```jldoctest
│ julia> C = combinations(4, 3)
│ Iterator over the 3-combinations of 1:4
│ 
│ julia> collect(C)
│ 4-element Vector{Combination{Int64}}:
│  [1, 2, 3]
│  [1, 2, 4]
│  [1, 3, 4]
│  [2, 3, 4]
│ ```
│ 
│ Subexpression:
│ 
│ C = combinations(4, 3)
│ 
│ Evaluated output:
│ 
│ Iterator over the 3-combinations of Base.OneTo(4)
│ 
│ Expected output:
│ 
│ Iterator over the 3-combinations of 1:4
```

You can fix this of course by simply adjusting the doctest. But I wonder if it might be better to instead change the printing code, as `Base.OneTo(4)` is something of a technical detail and a casual user may be confused by this output

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 25]
```diff
@@ -73,13 +73,17 @@ Base.length(C::Combinations) = binomial(C.n, C.k)
 
 Base.eltype(::Type{Combinations{T}}) where {T} = Combination{eltype(T)}
 
+function Base.show(io::IO, C::Combinations{Base.OneTo})
```
Comment (fingolfin) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: Oh, I see you actually already do have a custom `show` method, excellent! It just doesn't get used... Presumably because you need this
```suggestion
function Base.show(io::IO, C::Combinations{<:Base.OneTo})
```

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 26]
```diff
@@ -18,7 +18,7 @@ julia> collect(C)
  [2, 3, 4]
 ```
 """
-combinations(n::Int, k::Int) = Combinations(1:n, k)
+combinations(n::Int, k::Int) = Combinations(range(; length=n), k)
```
Comment (fingolfin) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: Maybe like this? That would make it easier to "find all places related to the `OneTo` optimization" via grepping.
```suggestion
combinations(n::Int, k::Int) = Combinations(Base.OneTo(n), k)
```

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 27]
```diff
@@ -92,24 +96,178 @@ function Base.show(io::IO, ::MIME"text/plain", P::Combination)
...
+  return length(data(C))
+end
+
+function Base.getindex(C::Combination, i::IntegerUnion)
+  return getindex(data(C), Int(i))
+end
+
+function Base.setindex!(C::Combination, x::IntegerUnion, i::IntegerUnion)
+  return setindex!(data(C), x, i)
+end
```
Comment (fingolfin) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: Is that really a good idea? Do we want people to be able to easily modify a `Combination` object? I would have expected those to be "essentially immutable" (as in: regular user interfaces don't allow modifications. Of course the user can always bypass these and mess with internals, but that's on them then)

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl (hunk 26 above)
Comment (fingolfin) [reply src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: Ah and this is also necessary for this to work in Julia 1.6

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl (hunk 27 above)
Comment (thofma) [reply src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: I would disallow it.

Comment (mjrodgers) [reply src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: I'll take this out, I agree it doesn't make sense. I think this is from the initial `Combination` PR (though I changed it to use `C` instead of `P` for the variable name).  `Permutation` also has a `setindex!` method.

**Comment by fingolfin** [comment]
Shouldn't we also have an entry for "combinations" in the "Enumerative combinatorics" section in the manual, say at https://docs.oscar-system.org/dev/Combinatorics/EnumerativeCombinatorics/combinations? Which could list the relevant docstrings and perhaps some extra examples. A place where the user could find out about the new "random access" to `combinations` objects.

Or if there is deemed not enough material for a separate section, then perhaps there could be an "Intro" section for "enumerative combinatorics" that had this material, plus anything else we deem relevant?

**Comment by fingolfin** [comment]
(this doesn't have to be in this PR, but it would be good to have anyway -- after all, functionality our users can not find is useless for them)

**Comment by mjrodgers** [comment]
> Shouldn't we also have an entry for "combinations" in the "Enumerative combinatorics" section in the manual

Agreed, I can add this in a follow-up PR. I think it will make sense to also allow `Combination` for unit ranges to use `T<:IntegerUnion` as for `Permutation`, so that users can use `Int8` or similar for performance reasons, so I can do these together maybe.

**Comment by mjrodgers** [comment]
This should be ready to merge now, assuming I didn't mess up any of the tests with the last batch of minor tweaks.

**Review by fingolfin — APPROVED** [review]
